Filtered out zero value Boundaries in ZIO Metrics Core.

Grryum opened the pull request:
This enhancement aims to improve the accuracy and reliability of metrics reporting by removing meaningless zero value boundaries that can potentially cause failures in backend metrics systems.

The motivation behind this merge request is to address the issue with what I faced in project `zio-metrics-connectors` (`InstrumentedSample`) where zero value boundaries, such as zero duration or zero count, may be inadvertently captured and reported as metrics. In certain backend metrics systems, zero values can lead to unexpected behavior or errors during ingestion or processing. By filtering out these zero value boundaries at the core level, we ensure that only meaningful and relevant metrics are reported, reducing the likelihood of failures or inconsistencies in the monitoring infrastructure, amount of the whole traffic and metrics size.

Now it is very easy to create Boundaries with unuseful zero upper bound. For example:

`Histogram.Boundaries.linear(0.0d, 10.0d, 11)`

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/zio/zio?pullRequest=8150) <br/>All committers have signed the CLA.

adamgfraser commented:
@Grryum Your example does not produce any zero upper bound:

```scala
import zio._
import zio.metrics._

object Example extends App {
  val boundaries = MetricKeyType.Histogram.Boundaries.linear(0.0d, 10.0d, 11)
  println(boundaries) 
}
```

```
Boundaries(Chunk(0.0,10.0,20.0,30.0,40.0,50.0,60.0,70.0,80.0,90.0,100.0,1.7976931348623157E308))
```

Grryum commented:
@adamgfraser The problem I reported about is with value `0.0` in printed chunk. Moreover, some metric instruments interprets the presence of this value in bounds as an error, for example Micrometer. I faced with it when was preparing ZIO Metrics integration with Micrometer https://github.com/zio/zio-metrics-connectors/pull/42

adamgfraser commented:
@Grryum Please confirm that this does not break any existing metrics metrics implementations.

Grryum commented:
@adamgfraser for my part, I can confirm that this does not break compatibility any metrics implementations.
This change only affects on final metric data encoding to string in different vendor-dependent format:
1) prometheus: https://github.com/zio/zio-metrics-connectors/blob/zio/series2.x/core/src/main/scala/zio/metrics/connectors/prometheus/PrometheusEncoder.scala#L95-L103
2) statsd https://github.com/zio/zio-metrics-connectors/blob/zio/series2.x/core/src/main/scala/zio/metrics/connectors/statsd/StatsdEncoder.scala#L53-L56

But as I mentioned earlier, boundary with zero value doesn't make sense for histogram because it should represents a upper boundary of bucket.

I also fixed one existed test and added a new specialized one.

adamgfraser commented:
@Grryum Can you please sign the CLA?

Grryum commented:
> @Grryum Can you please sign the CLA?

@adamgfraser I already have done it several times :( Can you please check now

adamgfraser commented:
@Grryum It looks like your first commit was from a different account.

Grryum commented:
> @Grryum It looks like your first commit was from a different account.

I have no idea how to fix it :( maybe you have?

Grryum committed:
Filtered out zero value in main Boundaries constructor method.

Grryum committed:
Fixed one test and added another for testing boundaries without zero and negative values.

Grryum commented:
@adamgfraser finally got win over CLA :)
